File: apps/client/src/components/paginated-list/PagingView.swift
internal import ExpoModulesCore
internal import ExpoUI
internal import VoelNativePaging

/// Props have the native view's lifetime, unlike transient SwiftUI body values.
class PagingViewProps<Value: Record & AnyObject>: UIBaseViewProps {
  @Field var paging: PagingOptions?
  var onPageRequest = EventDispatcher()
  var onPageCancel = EventDispatcher()
  private var currentPager: NativePager<Value>?

  // Sizing is creation-time configuration; the React wrapper remounts on changes.
  func pager(options: PagingOptions) -> NativePager<Value> {
    if let currentPager { return currentPager }
    let context = appContext
    let pager = NativePager<Value>(
      options: options,
      decode: {
        guard let context else { throw Exceptions.AppContextLost() }
        return try Value(from: $0, appContext: context)
      },
      onRequest: { [weak self] in self?.onPageRequest($0) },
      onCancel: { [weak self] in self?.onPageCancel($0) }
    )
    currentPager = pager
    return pager
  }

  func resolvePage(_ page: PageDelivery) throws { try currentPager?.resolve(page) }
  func rejectPage(_ request: PageCancellation) { currentPager?.reject(request) }
}

File: apps/client/src/components/paginated-list/PagingView.kt
package app

import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.remember
import androidx.compose.runtime.getValue
import app.voel.paging.NativePager
import app.voel.paging.PageCancellation
import app.voel.paging.PageDelivery
import app.voel.paging.PageRequestEvent
import app.voel.paging.PagingOptions
import expo.modules.kotlin.views.ComposeProps
import expo.modules.kotlin.views.ComposeViewBuilderScope
import expo.modules.kotlin.views.FunctionalComposableScope

interface PagingViewProps : ComposeProps {
    val paging: PagingOptions?
}

/** Installs private view commands/events and owns one pager per native composition. */
fun <Props : PagingViewProps, Value : Any> ComposeViewBuilderScope<Props>.PagingContent(
    decode: (Map<String, Any?>) -> Value,
    content: @Composable FunctionalComposableScope.(Props, NativePager<Value>) -> Unit,
) {
    val onPageRequest by Event<PageRequestEvent>()
    val onPageCancel by Event<PageCancellation>()
    val resolvePage by AsyncFunction<PageDelivery>()
    val rejectPage by AsyncFunction<PageCancellation>()

    Content { props ->
        val options = props.paging ?: return@Content
        val pager = remember(options.pageSize, options.maxResidentItems) {
            NativePager(options, decode, { onPageRequest(it) }, { onPageCancel(it) })
        }
        resolvePage.handle { pager.resolve(it) }
        rejectPage.handle { pager.reject(it) }
        DisposableEffect(pager) {
            pager.start()
            onDispose { pager.close() }
        }
        content(props, pager)
    }
}
